File: medium/739. Daily Temperatures/739. Daily Temperatures.py
from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        # TLE
        # res = []
        # for i in range(len(temperatures)):
        #     count = 0
        #     for j in range(i + 1, len(temperatures)):
        #         if temperatures[j] > temperatures[i]:
        #             count = j - i
        #             break
        #     res.append(count)
        # return res

        n = len(temperatures)
        res = [0] * n
        stack = []

        for i in range(n):
            while stack and temperatures[i] > temperatures[stack[-1]]:
                prev = stack.pop()
                res[prev] = i - prev

            stack.append(i)

        return res


solution = Solution()
print(solution.dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]))
print(solution.dailyTemperatures([30, 40, 50, 60]))
print(solution.dailyTemperatures([30, 60, 90]))

File: medium/739. Daily Temperatures/739. Daily Temperatures.go
package main

import "slices"

func dailyTemperatures(temperatures []int) []int {
	n := len(temperatures)
	res := slices.Repeat([]int{0}, n)
	stack := []int{}

	for i := 0; i < n; i++ {
		for len(stack) > 0 && temperatures[i] > temperatures[stack[len(stack)-1]] {
			prev := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			res[prev] = i - prev
		}

		stack = append(stack, i)
	}

	return res
}
